dist/gcc-compatible/Hacl_Streaming_HMAC.c

```diff
@@ -0,0 +1,2204 @@
+  if (scrut.tag == Blake2B_s)
+  {
+    uint64_t *p1 = scrut.case_Blake2B_s;
+    Hacl_Hash_Blake2b_finish(64U, dst, p1);
+    return;
+  }
+  if (scrut.tag == Blake2B_256_s)
+  {
+    Lib_IntVector_Intrinsics_vec256 *p1 = scrut.case_Blake2B_256_s;
+    #if HACL_CAN_COMPILE_VEC256
+    Hacl_Hash_Blake2b_Simd256_finish(64U, dst, p1);
+    return;
+    #else
+    KRML_MAYBE_UNUSED_VAR(p1);
+    return;
+    #endif
+  }
+  KRML_HOST_EPRINTF("KaRaMeL abort at %s:%d\n%s\n",
+    __FILE__,
+    __LINE__,
+    "unreachable (pattern matches are exhaustive in F*)");
+  KRML_HOST_EXIT(255U);
+}
+
+static void free_(Hacl_Agile_Hash_state_s *s)
+{
+  Hacl_Agile_Hash_state_s scrut = *s;
+  if (scrut.tag == MD5_s)
+  {
+    uint32_t *p1 = scrut.case_MD5_s;
+    KRML_HOST_FREE(p1);
+  }
+  else if (scrut.tag == SHA1_s)
+  {
+    uint32_t *p1 = scrut.case_SHA1_s;
+    KRML_HOST_FREE(p1);
+  }
+  else if (scrut.tag == SHA2_224_s)
+  {
+    uint32_t *p1 = scrut.case_SHA2_224_s;
+    KRML_HOST_FREE(p1);
+  }
+  else if (scrut.tag == SHA2_256_s)
+  {
+    uint32_t *p1 = scrut.case_SHA2_256_s;
+    KRML_HOST_FREE(p1);
+  }
+  else if (scrut.tag == SHA2_384_s)
+  {
+    uint64_t *p1 = scrut.case_SHA2_384_s;
+    KRML_HOST_FREE(p1);
+  }
+  else if (scrut.tag == SHA2_512_s)
+  {
+    uint64_t *p1 = scrut.case_SHA2_512_s;
+    KRML_HOST_FREE(p1);
+  }
+  else if (scrut.tag == SHA3_224_s)
+  {
+    uint64_t *p1 = scrut.case_SHA3_224_s;
+    KRML_HOST_FREE(p1);
+  }
+  else if (scrut.tag == SHA3_256_s)
+  {
+    uint64_t *p1 = scrut.case_SHA3_256_s;
+    KRML_HOST_FREE(p1);
+  }
+  else if (scrut.tag == SHA3_384_s)
+  {
+    uint64_t *p1 = scrut.case_SHA3_384_s;
+    KRML_HOST_FREE(p1);
+  }
+  else if (scrut.tag == SHA3_512_s)
+  {
+    uint64_t *p1 = scrut.case_SHA3_512_s;
+    KRML_HOST_FREE(p1);
+  }
+  else if (scrut.tag == Blake2S_s)
+  {
+    uint32_t *p1 = scrut.case_Blake2S_s;
+    KRML_HOST_FREE(p1);
+  }
+  else if (scrut.tag == Blake2S_128_s)
+  {
+    Lib_IntVector_Intrinsics_vec128 *p1 = scrut.case_Blake2S_128_s;
+    KRML_ALIGNED_FREE(p1);
+  }
+  else if (scrut.tag == Blake2B_s)
+  {
+    uint64_t *p1 = scrut.case_Blake2B_s;
+    KRML_HOST_FREE(p1);
+  }
+  else if (scrut.tag == Blake2B_256_s)
+  {
+    Lib_IntVector_Intrinsics_vec256 *p1 = scrut.case_Blake2B_256_s;
+    KRML_ALIGNED_FREE(p1);
+  }
+  else
+  {
+    KRML_HOST_EPRINTF("KaRaMeL abort at %s:%d\n%s\n",
+      __FILE__,
+      __LINE__,
+      "unreachable (pattern matches are exhaustive in F*)");
+    KRML_HOST_EXIT(255U);
+  }
+  KRML_HOST_FREE(s);
+}
+
+static void copy(Hacl_Agile_Hash_state_s *s_src, Hacl_Agile_Hash_state_s *s_dst)
+{
+  Hacl_Agile_Hash_state_s scrut = *s_src;
+  if (scrut.tag == MD5_s)
+  {
+    uint32_t *p_src = scrut.case_MD5_s;
+    Hacl_Agile_Hash_state_s x1 = *s_dst;
+    uint32_t *p_dst;
+    if (x1.tag == MD5_s)
+    {
+      p_dst = x1.case_MD5_s;
+    }
+    else
+    {
+      p_dst = KRML_EABORT(uint32_t *, "unreachable (pattern matches are exhaustive in F*)");
```

Comment:
Out of curiosity, but does the HACL* design require to always abort and not to return an error code in case of such failures?

Reply:
In the verified code, there is a precondition that says that the caller cannot mix-and-match for the copy function, and can only copy between states that have identical choices of implementation and key length. This means that from the point of view of the verified code, this simply cannot happen.

In reality, however, callers are in C, and the C compiler does not check that precondition, so this line of code could still be reachable. All we have is a comment (TODO) that says don't do that. So the code generation adds this line so that i) there is an actionable error rather than a segfault that requires firing up the debugger, and ii) to avoid compiler warnings along the lines of "this variable is possibly unset".

I could change the verified so that the code first checks at runtime that the choice of implementation and key lengths match, and returns an error otherwise -- as opposed to a precondition that the C code might miss. Thoughts?

Reply:
Looking at it more closely, my explanation above is wrong, because the code returns a fresh state for the copy, and this is a union selection, so the code won't pick the "wrong branch". (I though the copy function took both source and destination pointers.)

In that case, yes, this could be optimized too.

Reply:
(Note that this is an internal copy function, and that the user-facing Hacl_Streaming_HMAC_copy function returns a fresh state.)

Reply:
Is there a way for `Hacl_Streaming_HMAC_copy` to fail and return NULL without triggering a crash?

Reply:
> So unless I'm missing something, Hacl_Streaming_HMAC_copy cannot trigger an error

For instance, what happens if malloc() returns NULL? (for instance, we're out of memory). We do check when malloc() returns NULL on our side, but I wondered whether you considered this.

Reply:
Right, that we generally do not check for in HACL*. We could, if deeemed important. Let us know!

Reply:
I personally would like to be able to raise a `PyErr_NoMemory()` which would anyway lead to a MemoryError (and generally an interpreter abort but not always I think). So I would prefer raising that exception if possible (and it would make sense that the allocation function behaves like a dedicated malloc() with the same semantics, namely returning "NULL" = no memory or something else happened).

If it's too hard or could break things on your side, it's also fine. (Probably, it'd be good to get Gregory's opinion on this matter as well).